docs/design/node_agent_config-design.md

```diff
@@ -0,0 +1,91 @@
+# NodeAgentConfig configuration with restic/kopia
+Date: 2023-07-25
+
+## Abstract
+
+New configuration for the file system backup & restore that will be used by the OADP and allow to choose Restic or Kopia uploader type.
+
+## Background
+
+For the file system backup and restore the Velero may use Restic or Kopia as the uploader mechanism. A number of tests were [performed](https://velero.io/docs/main/performance-guidance/) by the Velero community to compare those mechanisms. In many cases Kopia uploader is a much better performing mechanism and as such should be added to the OADP operator.
+
+## Goals
+
+- A new option `nodeagentconfig` to allow configuration of restic or kopia uploader
+- Backwards compatibility with the current OADP configuration schema options, namely `restic`
+- Preparation for the future deprecation of the current `restic` configuration option (from OADP 1.4+)
+- Allow new schema to be used by the datamover node agent
+
+## Non Goals
+- Enablment of datamover node agent
+- Deprecation of the "restic" configuration option
+- E2E tests for the `kopia` or `restic` uplaoder
+
+## High-Level Design
+
+Since new `nodeagentconfig` configuration option is a sibling of the `restic` one, the `restic` will be used as part of the new `nodeagentconfig` with addition of one `UploaderType` option that will be either `kopia` or `restic`.
+
+Without that option a default `restic` will be used. That default option may change in the future.
+
+## Detailed Design
+
+A new structure will be added, that includes the ResticConfig inline and extends this with one new parameter `UploaderType`:
+
+```go
+type NodeAgentConfig struct {
+	// Embedding ResticConfig
+	// +kubebuilder:validation:Optional
+	ResticConfig `json:",inline"`
```

Comment:
leftover comma?

Reply:
I think the idea here is to minimize duplicate struct fields. When Restic is fully gone from oadp that is when fields from resticconfig will be copied over instead of inherited inline.

Reply:
@shawn-hurley @kaovilai That was my first concern as well. If we want, though, we could restructure this and create a new NodeAgentFields struct and then have *that* struct inlined by both ResticConfig and NodeAgentConfig. Should be backwards-compatible for 1.2 ResticConfig as long as the inlined fields are identical to the prior inherited fields. Would that work, or does it break any important design patterns here? The main goal is to not have to copy-and-paste every field from ResticConfig into the new struct.

Reply:
In any case, once we completely remove ResticConfig in a future release, we can just put all of these fields directly in NodeAgentConfig.

Reply:
That is what I was driving towards as well. I think the only thing I would do, is add 1 more struct that does not have UploaderType because that is a specific field for the new NodeAgentConfig

Reply:
yeah, it could work with validation, but I would just as soon not give it as an option to the user if there is only one right answer you know?

Reply:
Also, if UploaderType is optional, are we going to just use velero default for NodeAgentConfig (i.e. kopia)? Since this is a new struct, defaulting to Kopia here won't suprise anyone on upgrade, since upgraded DPAs will use ResticConfig. If we default  UploaderType to Restic here, we will have an upgrade problem in 1.4 if we make the default Kopia -- *unless* part of the design is for the controller to modify the NodeAgentConfig to fill in UploaderType when it's left blank. That solves the future default problem because upgrades with existing DPAs won't have an empty field to default.

Reply:
So my take on this is that it adds some lines and when deprecating ResticConfig one would also want to move NodeAgentCommonFields into NodeAgentConfig and remove NodeAgentCommonFields all together, which makes just extra struct that in fact is the ResticConfig struct. I may be missing something, but looks like artificial addition to me. 

In other words with deprecation of `ResticConfig` we will in the future just move the relevant fields into `NodeAgentConfig` similarly to what we will do with the `NodeAgentCommonFields`.

Reply:
We're deprecating ResticConfig right now (for 1.3). So we will want to add deprecation warning condition to the DPA on reconcile (but not an error, since it's still functional). Removing it entirely will be in a future release. 

We wouldn't have to get rid of the common fields struct when removing ResticConfig (although it would simplify things to do so). I guess my take is that there are more lines with the common struct, but the resulting structs are clearer in their purpose. Having an optional field in ResticConfig with only one valid value seems confusing from a user point of view. So the extra complication on the golang struct side of things results in a simpler yaml/CRD API.

Reply:
I don't know if I agree that we would want to add the NodeAgentCommonFields into NodeAgentConfig. It may be confusing, but I do think that in the future, we *may* have additional NodeAgents (for dealing with third-party DataMovers). It will be beneficial to have this common struct IMO.

Reply:
Ok I will add that common struct then. Thanks for all great comments and guiding me on this.

Reply:
Closing as we have common structure and I feel like all other considered options are covered in the "Alternatives Considered"